### PR TITLE
OTA-861: inhibit the 2nd minor version upgrade

### PR DESCRIPTION
**Pull request by hongkailiu**
This PR prevents the cluster to be upgraded to x.y+2.z1 while
the upgrade to x.y+1.z2 from x.y.z3 is still in progress.

### CODE REVIEW

**Comment by Davoska** [comment]
Hello @hongkailiu :wave:
The pull request contains two commits that have an identical message:
```shell
c9852051 (HEAD -> OTA-861) inhibit the 2nd minor version upgrade
08ece187 inhibit the 2nd minor version upgrade
```

Could you please update the commits and provide some additional information in them as well? The additional information would be greatly appreciated :raised_hands:

**Comment by petr-muller** [comment]
I'll sit this one out and leave it for David to review :wink:

**Comment by Davoska** [comment]
Petr, back to status API!

**Review by Davoska — CHANGES_REQUESTED** [review]
I would like to discuss the wording of the precondition check :raised_hands: 

Other than that, the PR seems to be good to merge.

**Comment by Davoska** [comment]
tldr; everything seems to be working as expected :raised_hands:
---

I wanted to learn something new by testing this PR. I have created three dummy releases. All these releases contain this PR and only differ in their version. I had to force the upgrades, as the releases were not signed. 

Upgrade from `4.17.0-0.ci.test-2024-08-26-132104-ci-ln-c7kz9p2-latest` to `4.18.98`:

```
$ oc adm upgrade --to-image quay.io/dhurta/ocp-release@sha256:bdb2e638b14cb5bfbc1e1502b983210dce43e6258827cedf65311391bc133065 --allow-explicit-upgrade --force --allow-upgrade-with-warnings 
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures.
Requested update to release image quay.io/dhurta/ocp-release@sha256:bdb2e638b14cb5bfbc1e1502b983210dce43e6258827cedf65311391bc133065
```

Checking events:

```
$ oc events -n openshift-cluster-version
...
4m39s                   Warning   PreconditionWarn      ClusterVersion/version                           precondition warning for payload loaded version="4.18.98" image="quay.io/dhurta/ocp-release@sha256:bdb2e638b14cb5bfbc1e1502b983210dce43e6258827cedf65311391bc133065": Precondition "ClusterVersionRecommendedUpdate" failed because of "NoChannel": Configured channel is unset, so the recommended status of updating from 4.17.0-0.ci.test-2024-08-26-150237-ci-ln-dxk8yqk-latest to 4.18.98 is unknown.
```

Only `ClusterVersionRecommendedUpdate` precondition check warning as expected.

---

Upgrade from `4.18.98` to `4.18.99`:

```
$ oc adm upgrade --to-image quay.io/dhurta/ocp-release@sha256:a570a6a0957729b3f876750f774c9bb2411fd3596093b587f2de998604cdd5ee --allow-explicit-upgrade --force --allow-upgrade-with-warnings 
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures.
warning: --allow-upgrade-with-warnings is bypassing: the cluster is already upgrading:

  Reason: ClusterOperatorsUpdating
  Message: Working towards 4.18.98: 110 of 900 done (12% complete), waiting on etcd, kube-apiserver
Requested update to release image quay.io/dhurta/ocp-release@sha256:a570a6a0957729b3f876750f774c9bb2411fd3596093b587f2de998604cdd5ee
```

Checking events:

```
$ oc events -n openshift-cluster-version
...
3m9s                    Warning   PreconditionWarn      ClusterVersion/version                           precondition warning for payload loaded version="4.18.99" image="quay.io/dhurta/ocp-release@sha256:a570a6a0957729b3f876750f774c9bb2411fd3596093b587f2de998604cdd5ee": Precondition "ClusterVersionRecommendedUpdate" failed because of "NoChannel": Configured channel is unset, so the recommended status of updating from 4.18.98 to 4.18.99 is unknown.
```

The new precondition check does not fail when redirecting a partial minor upgrade to a newer patch version in the same minor version.

---

Upgrade from `4.18.99` to `4.19.99`:

```
$ oc adm upgrade --to-image quay.io/dhurta/ocp-release@sha256:88a0a7c9e196b7df971c50856f5024891c8ecb2ce61a84220e2b193c3ced33f3 --allow-explicit-upgrade --force --allow-upgrade-with-warnings 
warning: The requested upgrade image is not one of the available updates. You have used --allow-explicit-upgrade for the update to proceed anyway
warning: --force overrides cluster verification of your supplied release image and waives any update precondition failures.
warning: --allow-upgrade-with-warnings is bypassing: the cluster is already upgrading:

  Reason: ClusterOperatorsUpdating
  Message: Working towards 4.18.99: 110 of 900 done (12% complete), waiting on etcd, kube-apiserver
Requested update to release image quay.io/dhurta/ocp-release@sha256:88a0a7c9e196b7df971c50856f5024891c8ecb2ce61a84220e2b193c3ced33f3
```

Checking events:

```
$ oc events -n openshift-cluster-version
...
83s                     Warning   PreconditionWarn      ClusterVersion/version                           precondition warning for payload loaded version="4.19.99" image="quay.io/dhurta/ocp-release@sha256:88a0a7c9e196b7df971c50856f5024891c8ecb2ce61a84220e2b193c3ced33f3": Forced through blocking failures: Multiple precondition checks failed:
* Precondition "ClusterVersionUpgradeable" failed because of "MinorVersionClusterUpgradeInProgress": The minor level upgrade to 4.19.99 is not allowed while there is already a minor level upgrade from 4.17.0-0.ci.test-2024-08-26-150237-ci-ln-dxk8yqk-latest to 4.18.99 in progress. Please wait until the existing minor level upgrade completes.
* Precondition "ClusterVersionRecommendedUpdate" failed because of "NoChannel": Configured channel is unset, so the recommended status of updating from 4.18.99 to 4.19.99 is unknown.
```

The precondition check `ClusterVersionUpgradeable` fails as expected because of the `MinorVersionClusterUpgradeInProgress` check.

**Comment by hongkailiu** [comment]
> I have created three dummy releases. All these releases contain this PR and only differ in their version.

**Comment by hongkailiu** [comment]
Thanks for the cmds creating a release with PR.
I will ask the testing strategy in our team channel.

**Comment by hongkailiu** [comment]
/hold

https://github.com/openshift/cluster-version-operator/pull/1080
